lib/src/merge.dart

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+/// Emits values from the source stream and [other] in any order as they arrive.
+///
+/// If the source stream is a broadcast stream, the result stream will be as
+/// well, regardless of the type of stream [other] is.
+StreamTransformer<T, T> merge<T>(Stream<T> other) => new _Merge<T>([other]);
+
+/// Emits values from the source stream and all streams in [others] in any order
+/// as they arrive.
+///
+/// If the source stream is a broadcast stream, the result stream will be as
+/// well, regardless of the types of streams in [others].
```

Comment:
In general I think it *can* be sensible to merge streams of mixed types so I want to support it. We have to pick a behavior and picking it based on the source stream seems the most sensible to me.

I think it's most common for a `StreamTransformer` to preserve the type of the source stream. That's what `new StreamTransformer.fromHandlers()` would do which we use elsewhere in this package. I suppose we could drop this part of the comment altogether and most users will assume it, but I thought it might be worthwhile to make it explicit since there are multiple streams in question here.